src/lib/Signer/CasperWallet.ts

```diff
@@ -0,0 +1,119 @@
+import BaseSigner from './BaseSigner';
+
+export default class CasperWallet extends BaseSigner {
+  public isCasperWallet = true;
+
+  // Consider move these flags to the base class
+  private connected: boolean;
+  private locked: boolean;
+
+  private casperWalletProvider: ReturnType<CasperWalletProvider>;
+
+  constructor(options?: CasperWalletProviderOptions) {
+    super();
+
+    this.connected = false;
+    this.locked = true;
+
+    this.casperWalletProvider = window.CasperWalletProvider!(options);
+
+    // Register event listeners
+    window.addEventListener(
+      CasperWalletEventTypes.Connected,
+      this.handleEvent.bind(this)
+    );
+    window.addEventListener(
+      CasperWalletEventTypes.Disconnected,
+      this.handleEvent.bind(this)
+    );
+    window.addEventListener(
+      CasperWalletEventTypes.ActiveKeyChanged,
+      this.handleEvent.bind(this)
+    );
+    window.addEventListener(
+      CasperWalletEventTypes.Locked,
+      this.handleEvent.bind(this)
+    );
+    window.addEventListener(
+      CasperWalletEventTypes.Unlocked,
+      this.handleEvent.bind(this)
+    );
+  }
+
+  private handleEvent(event: any) {
+    try {
+      const state: CasperWalletState = JSON.parse(event.detail);
+
+      if (state.isConnected !== this.connected) {
+        if (state.isConnected) {
+          this.emit('connected', this.activeAccount!);
+        } else {
+          // @ts-ignore
+          this.emit('disconnected');
+        }
+      }
+      if (state.activeKey && this.activeAccount !== state.activeKey) {
+        this.emit('accountChanged', state.activeKey);
+      }
+      if (this.locked !== state.isLocked) {
+        if (state.isLocked) {
+          // @ts-ignore
+          this.emit('locked');
+        } else {
+          // @ts-ignore
+          this.emit('unlocked');
+        }
+      }
```

Comment:
This will emit 3 different events while only one event is emitted by the Casper wallet which changes isConnected status, active key and locked status. IMO each event should have a separate handler